File: go/ambler.go
package ambler

import (
	"context"
	"fmt"
)

// Next represents the result of a node execution.
type Next[S any, K comparable] struct {
	NodeID *K
	State  S
}

// Node is a function that represents a node in the state machine.
// S is the state type, K is the node identifier type.
type Node[S any, K comparable] func(ctx context.Context, state S, key K) (Next[S, K], error)

// AmblerFunc is the type returned by the Ambler function.
type AmblerFunc[S any, K comparable] func(ctx context.Context, nodeID K, state S) (Next[S, K], error)

// Ambler creates a single-step executor for a node registry.
func Ambler[S any, K comparable](nodes map[K]Node[S, K]) AmblerFunc[S, K] {
	return func(ctx context.Context, nodeID K, state S) (Next[S, K], error) {
		node, ok := nodes[nodeID]
		if !ok {
			return Next[S, K]{}, fmt.Errorf("node not found: %v", nodeID)
		}

		return node(ctx, state, nodeID)
	}
}

// Ptr is a helper to return a pointer to a value, useful for NodeID in Next.
func Ptr[T any](v T) *T {
	return &v
}

File: python/ambler.py
from typing import TypeVar, Union, Callable, Dict, Tuple, Optional, Any, Awaitable
import asyncio

S = TypeVar("S")
K = TypeVar("K", bound=str)
H = TypeVar("H", bound=str)

T = TypeVar("T")

# Represents a value that can be either a synchronous value or an Awaitable of that value.
MaybeAwaitable = Union[T, Awaitable[T]]

# A map from edge names to the next node identifier (or None to terminate).
Edges = Dict[H, Optional[K]]

# The result returned by a node: a tuple of [nextNodeId, newState].
Next = Tuple[Optional[K], S]

# A function that represents a node in the state machine.
Node = Callable[[S, K], MaybeAwaitable[Next[S, K]]]

def ambler(nodes: Dict[K, Node[S, K]]) -> Callable[[K, S], MaybeAwaitable[Next[S, K]]]:
    """
    Creates a single-step executor for a node registry.
    """
    def execute(node_id: K, state: S) -> MaybeAwaitable[Next[S, K]]:
        node = nodes.get(node_id)
        if not node:
            raise ValueError(f"Node not found: {node_id}")

        return node(state, node_id)
    
    return execute
